AttributeKeyProvider.php

```diff
@@ -0,0 +1,256 @@
+    ) {
+        $this->channelRepository = $channelRepository;
+        $this->localeRepository = $localeRepository;
+        $this->currencyRepository = $currencyRepository;
+    }
+
+    /**
+     * Generate the list of attribute keys for a given attribute
+     *
+     * Example:
+     * Attribute: description localizable in english and french for ecommerce
+     * Attributes keys are:
+     * [ description-en_US-ecommerce, description-fr_FR-ecommerce ]
+     *
+     * @param AttributeInterface $attribute
+     *
+     * @return array
+     */
+    public function getAttributeKeys(AttributeInterface $attribute)
+    {
+        $keys = [];
+
+        $keys[$attribute->getCode()] = [];
+
+        $updatedKeys = [];
+        if ($attribute->isScopable() && $attribute->isLocalizable()) {
+            foreach ($this->getLocales() as $locale) {
+                foreach ($this->getChannels() as $channel) {
+                    foreach ($keys as $baseKey => $keyOptions) {
+                        $key               = $baseKey . '-' . $locale->getCode() . '-' . $channel->getCode();
+                        $updatedKeys[$key] = array_merge($keyOptions, ['locale' => $locale, 'channel' => $channel]);
+                    }
+                }
+            }
+            $keys = $updatedKeys;
+        } elseif ($attribute->isScopable() && !$attribute->isLocalizable()) {
+            foreach ($this->getChannels() as $channel) {
+                foreach ($keys as $baseKey => $keyOptions) {
+                    $key               = $baseKey . '-' . $channel->getCode();
+                    $updatedKeys[$key] = array_merge($keyOptions, ['channel' => $channel]);
+                }
+            }
+            $keys = $updatedKeys;
+        } elseif (!$attribute->isScopable() && $attribute->isLocalizable()) {
+            foreach ($this->getLocales() as $locale) {
+                foreach ($keys as $baseKey => $keyOptions) {
+                    $key               = $baseKey . '-' . $locale->getCode();
+                    $updatedKeys[$key] = array_merge($keyOptions, ['locale' => $locale]);
+                }
+            }
+            $keys = $updatedKeys;
+        }
+
+        switch ($attribute->getBackendType()) {
+            case 'prices':
+                $updatedKeys = [];
+
+                foreach ($keys as $key => $keyOptions) {
+                    foreach ($this->getCurrencies() as $currency) {
+                        $updatedKeys[$key . '-' . $currency->getCode()] = array_merge(
+                            $keyOptions,
+                            ['currency' => $currency]
+                        );
+                    }
+                }
+                $keys = $updatedKeys;
+                break;
+            case 'metric':
+                $updatedKeys = [];
+
+                foreach ($keys as $key => $keyOptions) {
+                    $updatedKeys[$key]                           = $keyOptions;
```

Comment:
A nice illustration of the inconvenient of aligning `=` everywhere ;) The align code is worst than without alignment.